WalletWasabi.Gui/Tabs/EncryptionManager/DecryptMessageViewModel.cs

```diff
@@ -0,0 +1,221 @@
+﻿using Avalonia.Diagnostics.ViewModels;
+using Avalonia.Threading;
+using NBitcoin;
+using ReactiveUI;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Reactive.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Gui.Tabs.EncryptionManager;
+using WalletWasabi.Gui.Tabs.WalletManager;
+using WalletWasabi.Gui.ViewModels;
+using WalletWasabi.Helpers;
+using WalletWasabi.KeyManagement;
+
+namespace WalletWasabi.Gui.Tabs.EncryptionManager
+{
+	internal class DecryptMessageViewModel : CategoryViewModel, IDisposable
+	{
+		private string _encryptedMessage;
+		private string _password;
+		private string _decryptedMessage;
+		private string _myPublicKey;
+		private string _warningMessage;
+		private bool _isPublicKeyPresent;
+		private bool _isSearchResultExpanded;
+		private ObservableCollection<AddressPubKeyViewModel> _addresses;
+		private AddressPubKeyViewModel _selectedItem;
+		private IEnumerable<AddressPubKeyViewModel> _allKeysViewModels;
+		private string _addressSearch;
+		private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+
+		public string EncryptedMessage
+		{
+			get => _encryptedMessage;
+			set => this.RaiseAndSetIfChanged(ref _encryptedMessage, value);
+		}
+
+		public string Password
+		{
+			get => _password;
+			set => this.RaiseAndSetIfChanged(ref _password, value);
+		}
+
+		public string MyPublicKey
+		{
+			get => _myPublicKey;
+			set => this.RaiseAndSetIfChanged(ref _myPublicKey, value);
+		}
+
+		public string DecryptedMessage
+		{
+			get => _decryptedMessage;
+			set => this.RaiseAndSetIfChanged(ref _decryptedMessage, value);
+		}
+
+		public string AddressSearch
+		{
+			get => _addressSearch;
+			set => this.RaiseAndSetIfChanged(ref _addressSearch, value);
+		}
+
+		public string WarningMessage
+		{
+			get => _warningMessage;
+			set => this.RaiseAndSetIfChanged(ref _warningMessage, value);
+		}
+
+		public bool IsPublicKeyPresent
+		{
+			get => _isPublicKeyPresent;
+			set => this.RaiseAndSetIfChanged(ref _isPublicKeyPresent, value);
+		}
+
+		public bool IsSearchResultExpanded
+		{
+			get => _isSearchResultExpanded;
+			set => this.RaiseAndSetIfChanged(ref _isSearchResultExpanded, value);
+		}
+
+		public AddressPubKeyViewModel SelectedItem
+		{
+			get => _selectedItem;
+			set => this.RaiseAndSetIfChanged(ref _selectedItem, value);
+		}
+
+		public ObservableCollection<AddressPubKeyViewModel> Addresses
+		{
+			get => _addresses;
+			set => this.RaiseAndSetIfChanged(ref _addresses, value);
+		}
+
+		public ReactiveCommand DecryptCommand { get; }
+		public ReactiveCommand MyPublicKeyCommand { get; }
+
+		public EncryptionManagerViewModel Owner { get; }
+
+		public DecryptMessageViewModel(EncryptionManagerViewModel owner) : base("Decrypt Message")
+		{
+			Owner = owner;
+
+			var canDecrypt = this.WhenAnyValue(x => x.EncryptedMessage,
```

Comment:
I think it could be okay to check the public key is not empty too because you need both (message and public key) in order to be able to decrypt

Reply:
fixed